openhands/runtime/modal/runtime.py

```diff
@@ -0,0 +1,347 @@
+        if runtime_container_image_id:
+            return modal.Image.from_registry(runtime_container_image_id).run_commands(
+                """
+# Disable bracketed paste
+# https://github.com/pexpect/pexpect/issues/669
+echo "set enable-bracketed-paste off" >> /etc/inputrc && \\
+echo 'export INPUTRC=/etc/inputrc' >> /etc/bash.bashrc
+                """.strip()
+            )
+        elif base_container_image_id:
+            if 'ubuntu' in base_container_image_id and (
+                base_container_image_id.endswith(':latest')
+                or base_container_image_id.endswith(':24.04')
+            ):
+                libgl_package = 'libgl1'
+            else:
+                libgl_package = 'libgl1-mesa-glx'
+
+            extra_deps_line = ''
+            if runtime_extra_deps:
+                extra_deps_line = f'\n{runtime_extra_deps} && \\'
+
+            # Build the source into a tarball so we can hash it efficiently
+            # and avoid rebuilding the image unless the source changes.
+            self.build_dir_handler = tempfile.TemporaryDirectory()
+            build_dir = self.build_dir_handler.__enter__()
+
+            tarball_path, package_version = put_source_tarball_to_dir(build_dir)
+
+            # Commands lifted from `Dockerfile.j2`
+            return (
+                modal.Image.from_registry(base_container_image_id)
+                .apt_install(
+                    'wget',
+                    'sudo',
+                    'apt-utils',
+                    libgl_package,
+                    'libasound2-plugins',
+                    'git',
+                )
+                # Remove UID 1000 if it's called pn--this fixes the nikolaik image for ubuntu users
+                .run_commands(
+                    'if getent passwd 1000 | grep -q pn; then userdel pn; fi',
+                    'mkdir -p /openhands/logs && mkdir -p /openhands/poetry',
+                )
+                .env({'POETRY_VIRTUALENVS_PATH': '/openhands/poetry'})
+                .run_commands(
+                    """
+if [ ! -d /openhands/miniforge3 ]; then \
+wget --progress=bar:force -O Miniforge3.sh "https://github.com/conda-forge/miniforge/releases/latest/download/Miniforge3-$(uname)-$(uname -m).sh" && \
+bash Miniforge3.sh -b -p /openhands/miniforge3 && \
+rm Miniforge3.sh && \
+chmod -R g+w /openhands/miniforge3 && \
+bash -c ". /openhands/miniforge3/etc/profile.d/conda.sh && conda config --set changeps1 False && conda config --append channels conda-forge"; \
+fi
+""".strip(),
+                    '/openhands/miniforge3/bin/mamba install conda-forge::poetry python=3.11 -y',
+                )
+                .copy_local_file(tarball_path, '/openhands/openhands_ai.tar.gz')
+                .run_commands(
+                    f'cd /openhands && tar -xzf openhands_ai.tar.gz && mv openhands_ai-{package_version} code'
+                )
+                .workdir('/openhands/code')
+                .run_commands(
```

Comment:
We have some updates with our Dockerfile.j2 - is there a way to automatically sync those update to the modal box so we are able to always use the most up-to-date version?

Reply:
You could use `from_dockerfile`:
https://modal.com/docs/reference/modal.Image#from_dockerfile

This would be easier to keep in sync with `Dockerfile.j2`; the only downside is that the caching is worse. If you use the modal-specific syntax, you get caching on a per-layer basis, but if you use the Dockerfile syntax, if there's a change to the code, you have to rebuild the whole image. This may be worth it from a complexity standpoint; let me know which approach you prefer.

FWIW I think the best path in general is to use `runtime_container_image` so you don't have to worry about this.